crates/environ/src/fact/signature.rs

```diff
@@ -25,8 +27,25 @@ impl ComponentTypesBuilder {
     /// imported (matching whatever was `canon lift`'d) and functions that are
     /// exported (matching the generated function from `canon lower`).
     pub(super) fn signature(&self, options: &AdapterOptions, context: Context) -> Signature {
+        let mem_opts = match &options.options.data_model {
+            DataModel::LinearMemory(mem_opts) => mem_opts,
+            DataModel::Gc {} => {
+                match &self.module_types_builder()[options.options.core_type]
+                    .composite_type
+                    .inner
+                {
+                    crate::WasmCompositeInnerType::Func(f) => {
+                        return Signature {
+                            params: f.params().iter().map(|ty| self.val_type(ty)).collect(),
+                            results: f.returns().iter().map(|ty| self.val_type(ty)).collect(),
+                        };
+                    }
+                    _ => unreachable!(),
+                }
```

Comment:
Would it be possible to use this branch independent of whatever data model is used now?

Reply:
Ah I see what you were getting at this whole time now. Yes, done.